pallets/contracts/src/lib.rs

```diff
@@ -694,4 +777,28 @@ where
             Err(error) => Err(DispatchErrorWithPostInfo { post_info, error }),
         }
     }
+
+    fn base_upgrade_api(
+        origin: T::RuntimeOrigin,
+        contract_id: T::AccountId,
+        api: Api,
+        chain_version: ChainVersion,
+        api_code_hash: ApiCodeHash<T>,
+    ) -> DispatchResult {
+        ensure_root(origin)?;
+
+        SupportedApiUpgrades::<T>::try_mutate(&contract_id, &api, |supported_upgrades| {
+            supported_upgrades
+                .try_insert(chain_version.clone(), api_code_hash.clone())
+                .map_err(|_| Error::<T>::MaxNumberOfApiUpgradesExceeded)
+        })?;
```

Comment:
If the list of upgrades for one `Api` hits the maximum length.  Then we need to prune older upgrades.  Most of the time we should only need two upgrade code hashes per Api.  When we do major breaking Polymesh upgrades, we need to pre-deploy a new Api upgrade before the chain runtime upgrade, so that contract will continue to work before and after the chain upgrade.